File: js/complete/no037.js
#!/usr/bin/env node

/* The number 3797 has an interesting property. Being prime itself,
   it is possible to continuously remove digits from left to right,
   and remain prime at each stage:
   3797, 797, 97, and 7.

   Similarly we can work from right to left: 3797, 379, 37, and 3.

   Find the sum of the only eleven primes that are both truncatable from
   left to right and right to left.

   NOTE: 2, 3, 5, and 7 are not considered to be truncatable primes. */

var fns = require('../functions.js'),
    operator = require('../operator.js'),
    timer = require('../timer.js');

function truncatedList(n, fromLeft) {
  function numJoinMap(arr) {
    return Number(arr.join(''));
  };

  var digs = n.toString().split('');
  if (fromLeft) {
    function subMapLeft(i) {
      return digs.slice(i);
    };
    return operator.range(digs.length).map(subMapLeft).map(numJoinMap);
  } else {
    // If the bool fromLeft is false, we are right
    function subMapRight(i) {
      return digs.slice(0, i + 1);
    };
    return operator.range(digs.length).map(subMapRight).map(numJoinMap);
  }
};

function truncatedAll(n) {
  var result = truncatedList(n, true).concat(truncatedList(n, false));
  return operator.uniq(result);
};

function isTruncatablePrime(n, primes) {
  var candidates = truncatedAll(n);
  for (var i = 0, candidate; candidate = candidates[i]; i++) {
    if (operator.inArray(candidate, primes) != -1) {
      continue;
    } else if (fns.isPrime(candidate)) {
      primes.push(candidate);
    } else {
      return false;
    }
  }
  return true;
};

function findFirstNTruncatable(n, maxN) {
  var result = [],
      // We don't include 2, 3, 5, or 7
      primes = fns.sieve(maxN).slice(4);
  for (var i = 0, prime; prime = primes[i]; i++) {
    if (isTruncatablePrime(prime, primes)) {
      result.push(prime);
    }

    if (result.length == n) {
      return result;
    }
  }

  if (result.length < n) {
    return; // Not enough found, raise maxN
  }
  return result;
};

exports.main = function(verbose) {
  if (typeof verbose == 'undefined') {
    verbose = false;
  }
  var ans = findFirstNTruncatable(11, Math.pow(10, 6));

  if (verbose) {
    return [operator.sum(ans), '.\nThe primes are: ',
            ans.join(', '), '.'].join('');
  } else {
    return operator.sum(ans);
  }
};

if (require.main === module) {
  timer.timer(37, exports.main);
}

File: python/complete/no037.py
#!/usr/bin/env python

# The number 3797 has an interesting property. Being prime itself,
# it is possible to continuously remove digits from left to right,
# and remain prime at each stage:
# 3797, 797, 97, and 7.

# Similarly we can work from right to left: 3797, 379, 37, and 3.

# Find the sum of the only eleven primes that are both truncatable from
# left to right and right to left.

# NOTE: 2, 3, 5, and 7 are not considered to be truncatable primes.

import copy

from python.decorators import euler_timer
from python.functions import is_prime
from python.functions import sieve

def truncated_list(n, from_left):
    if from_left:
        digs = [dig for dig in str(n)]
        return [int("".join(digs[i:])) for i in range(len(digs))]
    # If the bool from_left is false, we are right
    else:
        digs = [dig for dig in str(n)]
        return [int("".join(digs[:i + 1])) for i in range(len(digs))]

def truncated_all(n):
    return list(set(truncated_list(n, True) + truncated_list(n, False)))

def is_truncatable_prime(n, primes):
    candidates = truncated_all(n)
    for candidate in candidates:
        if candidate in primes:
            continue
        elif is_prime(candidate):
            primes.add(candidate)
        else:
            return False
    return True

def find_first_n_truncatable(n, max_n):
    result = []
    primes = set(sieve(max_n)[4:]) # We don't include 2, 3, 5, or 7
    for prime in copy.copy(primes):
        if is_truncatable_prime(prime, primes):
            result.append(prime)
        if len(result) == n:
            return result

    if len(result) < n:
        raise Exception("Not enough found, raise max_n")

    return result

def main(verbose=False):
    ans = find_first_n_truncatable(11, 10**6)

    if verbose:
        return "%s.\nThe primes are: %s." % (
            sum(ans), ", ".join(str(prime) for prime in ans))
    else:
        return sum(ans)

if __name__ == '__main__':
    print euler_timer(37)(main)(verbose=True)
